examples/exex/rollup/src/db.rs

```diff
@@ -0,0 +1,280 @@
+) -> eyre::Result<Option<AccountInfo>> {
+    match connection.deref().query_row::<String, _, _>(
+        "SELECT data FROM account WHERE address = ?",
+        (address.to_string(),),
+        |row| row.get(0),
+    ) {
+        Ok(account_info) => Ok(Some(serde_json::from_str(&account_info)?)),
+        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
+        Err(e) => Err(e.into()),
+    }
+}
+
+/// Get all storages for the provided address using the database connection. Connection can be
+/// either [rusqlite::Transaction] or [rusqlite::Connection].
+fn get_storages<C: Deref<Target = Connection>>(
+    connection: &C,
+    address: Address,
+) -> eyre::Result<Vec<(B256, U256)>> {
+    connection
+        .prepare("SELECT key, data FROM storage WHERE address = ?")?
+        .query((address.to_string(),))?
+        .mapped(|row| {
+            Ok((
+                B256::from_str(row.get_ref(0)?.as_str()?),
+                U256::from_str(row.get_ref(1)?.as_str()?),
+            ))
+        })
+        .map(|result| {
+            let (key, data) = result?;
+            Ok((key?, data?))
+        })
+        .collect()
+}
+
+impl reth::revm::Database for Database {
+    type Error = ProviderError;
+
+    fn basic(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        self.get_account(address).map_err(|_| ProviderError::UnsupportedProvider)
+    }
+
+    fn code_by_hash(&mut self, code_hash: B256) -> Result<Bytecode, Self::Error> {
+        let bytecode = self.connection().query_row::<String, _, _>(
+            "SELECT data FROM bytecode WHERE hash = ?",
+            (code_hash.to_string(),),
+            |row| row.get(0),
+        );
+        match bytecode {
+            Ok(data) => Ok(Bytecode::new_raw(Bytes::from_str(&data).unwrap())),
+            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(Bytecode::default()),
+            Err(_) => Err(ProviderError::UnsupportedProvider),
+        }
+    }
+
+    fn storage(&mut self, address: Address, index: U256) -> Result<U256, Self::Error> {
+        let storage = self.connection().query_row::<String, _, _>(
+            "SELECT data FROM storage WHERE address = ? AND index = ?",
+            (address.to_string(), index.to_string()),
+            |row| row.get(0),
+        );
+        match storage {
+            Ok(data) => Ok(U256::from_str(&data).unwrap()),
+            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(U256::ZERO),
+            Err(_) => Err(ProviderError::UnsupportedProvider),
+        }
+    }
+
+    fn block_hash(&mut self, number: U256) -> Result<B256, Self::Error> {
+        let block_hash = self.connection().query_row::<String, _, _>(
+            "SELECT hash FROM block WHERE number = ?",
+            (number.to_string(),),
+            |row| row.get(0),
+        );
+        match block_hash {
+            Ok(data) => Ok(B256::from_str(&data).unwrap()),
+            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(B256::default()),
```

Comment:
this should be a hard error since revm already performs bound checks

Reply:
right, fixed